src/main/java/dev/jbang/source/ScriptSource.java

```diff
@@ -86,6 +101,181 @@ private ScriptSource(ResourceRef resourceRef, String content) {
 		this.script = content;
 	}
 
+	public List<String> getCompileOptions() {
+		return collectOptions("JAVAC_OPTIONS");
+	}
+
+	protected String getCompilerBinary(String requestedJavaVersion) {
+		return resolveInJavaHome("javac", requestedJavaVersion);
+	}
+
+	protected Predicate<ClassInfo> getMainFinder() {
+		return pubClass -> pubClass.method("main", BaseBuildCommand.STRINGARRAYTYPE) != null;
+	}
+
+	protected String getExtension() {
+		return ".java";
+	}
+
+	// build with javac and then jar... todo: split up in more testable chunks
+	public IntegrationResult buildJar(RunContext ctx, File tmpJarDir, File outjar,
```

Comment:
Why is this refactor necessary? The `ScriptSource` is meant as a data class to hold all the information that could be extracted from a source file, it shouldn't contain code that "does stuff".

Reply:
because this process is different for kotlin projects (or any other lang, e.g., scala) and static methods are fiendishly difficult/impossible specialize whereas with inheritance you just need a few simple overloads, like i did in this diff, to customize things while reusing the bulk of the code.  Static methods, in general, make for lousy extensibility.

Reply:
yes - we should just have a separate set of classes/hiearchy for the tools. those can do what is needed.